# <div align="center"> SPECIAL TOPICS III </div>
## <div align="center"> Data Science for Social Scientists  </div>
### <div align="center"> ECO 4199 </div>
# <div align="center"> Assignment 2 </div>
## <div align="center" style="color:red">Charie Brady</div>
## <div align="center" style="color:red">300043672</div>

# INSTRUCTIONS
This Jupyter Notebook contains __3 Parts__ for a total of __10 points__. Some, __but not all__, of the resources needed to answer these questions are coming from what we have seen during the first 5 lectures. You may otherwise need to look for the answers using the internet. I highly recommend looking at the answers in [Stackoverflow](https://stackoverflow.com/) - these are usually the first answers referenced by Google. And of course you have my office hours!

Also, I sometimes left the output of the correction to make clear what I expect you to do.

<br>
Once you filled out all the tasks, you should restart your Kernel and use the <i>run all cells</i> command. Then you can save the Jupyter Notebook as <strong>LASTNAME_FIRSTNAME_A2.ipynb</strong> and post it in the assignment tab in Brightspace. 
<br>

The deadline for the submitting is <strong>November 2, 2021 at 11:59pm</strong>. 


<br>
In the spirit of open source softwares you are allowed to work collaboratively <strong> BUT </strong>:

- I shall receive one notebook per student (No notebook = zero).
- Coding should reflect personal work (Identical notebooks = zero). This includes but is not limited to:
    - your own name and student number (obviously)
    - your own comments on code 
    - you own names for variables, functions, dictionaries, dataframes and so on.
    
Have fun!


## Part 1 - Simple linear regression model

- Remember the Titanic dataset you used for Assignment 1?
- Now that you performed some data cleaning and visualization let's try to use it for statistical analysis
- Import the relevant packages (this was done for you)
- Save the path to the Titanic dataset on your laptop (or Drive) 
    - It means I should be able to change it to my own path and then run your entire notebook!
    - Failure to respect this will make you lose points
- Read the data into df, create a new variable named nameLength and show the head of the dataframe


- Using the pacakge [Statsmodels](https://www.statsmodels.org/stable/example_formulas.html) and OLS regression produce 4 regressions by regressing `Fare` on:
    1. `nameLength`
    2. `Age`
    3. `Parents/Children Aboard`, you may want to rename it first
    4. `Pclass`, careful you will need to specify this is a [categorical variable](https://www.statsmodels.org/stable/example_formulas.html#:~:text=is%20correctly%20specified.-,Categorical%20variables,%C2%B6,-Looking%20at%20the)
 
 
- For each regression add below a `Markdown` cell in which you will:
    1. Interpret the coefficient values
    2. Interpret their significance
    

- Finally, show that $\hat{\beta}_1$ estimated using the OLS method can also be obtained by computing the covariance of Y and X over the variance of X for your first model: `Fare ~ nameLength`

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import statsmodels.formula.api as smf
import sklearn.linear_model as skl_lm

from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error

In [2]:
titanicPath="C:/Users/16136/Desktop/Fall 2021/ECO4199"

In [3]:
df=pd.read_csv('/'.join([titanicPath,'titanic.csv']))
df['nameLength'] = df['Name'].str.len()
df.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  \
0    male  22.0                        1                        0   7.2500   
1  female  38.0                        1                        0  71.2833   
2  female  26.0                        0                        0   7.9250   
3  female  35.0                        1                        0  53.1000   
4    male  35.0                        0                        0   8.0500   

   nameLength  
0          22  
1          50  
2          21  
3          43  
4          23

### Model 1

In [4]:
mod1 = smf.ols('Fare ~ nameLength', data=df).fit()
print(mod1.summary())

                            OLS Regression Results                            
Dep. Variable:                   Fare   R-squared:                       0.030
Model:                            OLS   Adj. R-squared:                  0.029
Method:                 Least Squares   F-statistic:                     27.33
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           2.14e-07
Time:                        19:37:03   Log-Likelihood:                -4710.7
No. Observations:                 887   AIC:                             9425.
Df Residuals:                     885   BIC:                             9435.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.7054      4.986      1.546      0.1

Interpretation:

#### 1. Interpret the coefficient values

The intercept $\hat{\beta}_0 = 7.7054$ is not especially meaningful. It is the value when name length is zero, but nobody has a name of zero length. A realistic minimum name length would possibly be around 11-12 (4-5 digits for title, 3 letter first name, 1 digit space, 3 letter last name), so the intercept cannot be interpreted as the fare in the absence of name length.

Since $\hat{\beta}_1 = 0.9715$, a one digit increase in name length increases the fare price by ~97 cents. This is a positive relationship between name length and fare price, such that people with longer names typically paid larger fares.

#### 2. Interpret their significance

Given that $\rho = 0 < \alpha = 0.05$, the association between name length and fare is statistically significant at the 95% confidence level. 

####  $\hat{\beta}_1 = \frac{cov(Y,X)}{var(X)}$

In [5]:
Syx = df[['Fare','nameLength']].cov().iloc[0,1] # covariance
Vx = df['nameLength'].var()
beta_1 = Syx/Vx
print(f"The value for the slope obtained using OLS: {mod1.params['nameLength']} is the same as the one obtained from computing the covariance of X and Y over the variance of X: {beta_1}" )
      

The value for the slope obtained using OLS: 0.971470580149084 is the same as the one obtained from computing the covariance of X and Y over the variance of X: 0.9714705801490834


### Model 2

In [6]:
mod2 = smf.ols('Fare ~ Age', data=df).fit()
print(mod2.summary())

                            OLS Regression Results                            
Dep. Variable:                   Fare   R-squared:                       0.013
Model:                            OLS   Adj. R-squared:                  0.012
Method:                 Least Squares   F-statistic:                     11.31
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           0.000804
Time:                        19:37:03   Log-Likelihood:                -4718.6
No. Observations:                 887   AIC:                             9441.
Df Residuals:                     885   BIC:                             9451.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     20.6354      3.848      5.363      0.0

Interpretation:


#### 1. Interpret the coefficient values

The intercept $\hat{\beta}_0 = 20.6354$ is the predicted fare price when the passenger's age is zero. Again, this is not especially meaningful in this case given that infants do not typically have their own fare rates that are different than a child, teen, etc. 

Since $\hat{\beta}_1 = 0.396$, a one year increase in age is associated with an increase in the fare price by ~40 cents. This is a positive relationship between age and fare price, such that older passengers typically paid larger fares than younger passengers.

#### 2. Interpret their significance

Given that $\rho = 0.001 < \alpha = 0.05$, the association between age and fare is statistically significant at the 95% confidence level. This does not mean that the effect is large, only that it is statistically different from zero. 

### Model 3

In [7]:
df['Family_Aboard'] = df['Parents/Children Aboard']
mod3 = smf.ols('Fare ~ Family_Aboard', data=df).fit()
print(mod3.summary())

                            OLS Regression Results                            
Dep. Variable:                   Fare   R-squared:                       0.046
Model:                            OLS   Adj. R-squared:                  0.045
Method:                 Least Squares   F-statistic:                     43.09
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           8.90e-11
Time:                        19:37:03   Log-Likelihood:                -4703.1
No. Observations:                 887   AIC:                             9410.
Df Residuals:                     885   BIC:                             9420.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept        27.2134      1.808     15.051

Interpretation:


#### 1. Interpret the coefficient values

The intercept $\hat{\beta}_0 = 27.2134$ is the predicted fare price when the passenger has no children or parents aboard. In this case, the intercept is meaningful, since a zero value for the independent variable is a sensible value with respect to fare price.

Since $\hat{\beta}_1 = 13.2842$, each additional child or parent the passenger has aboard is associated with a $13.28 increase in fare (i.e., a positive relationship).

#### 2. Interpret their significance

Given that $\rho = 0 < \alpha = 0.05$, the association between name age and fare is statistically significant at the 95% confidence level.  

### Model 4

In [8]:
df['Pclass'] = df['Pclass'].astype("category")

mod4 = smf.ols('Fare ~ Pclass', data=df).fit()
print(mod4.summary())

                            OLS Regression Results                            
Dep. Variable:                   Fare   R-squared:                       0.353
Model:                            OLS   Adj. R-squared:                  0.351
Method:                 Least Squares   F-statistic:                     240.7
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           3.62e-84
Time:                        19:37:03   Log-Likelihood:                -4531.4
No. Observations:                 887   AIC:                             9069.
Df Residuals:                     884   BIC:                             9083.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      84.1547      2.729     30.841      

Interpretation:

#### 1. Interpret the coefficient values

The intercept $\hat{\beta}_0 = 84.1547$ is the predicted fare price for a passenger in first class.

The coefficient $\hat{\beta}_1 = -63.4925$ is the difference in fare price between a second class passenger and a first class passenger. It is negative, meaning second class passengers have lower fares. A second class passenger's predicted fare price is $\hat{\beta}_0 + \hat{\beta}_1$. 

The coefficient $\hat{\beta}_2 = -70.447$ is the difference in fare price between a third class passenger and a first class passenger. It is negative, meaning third class passengers have lower fares, which is also lower than second class passengers. A third class passenger's predicted fare price is $\hat{\beta}_0 + \hat{\beta}_2$. 

#### 2. Interpret their significance

Given that $\rho_1 = \rho_2 = 0 < \alpha = 0.05$, fare prices of second and third class passengers are significantly different than first class passengers at the 95% confidence level. This conclusion is also supported by the low p-value of the F-statistic, which tests whether the coefficients are jointly zero. 

## Part 2 - Multiple linear regression model


- Using the pacakge [Statsmodels](https://www.statsmodels.org/stable/example_formulas.html) and OLS regression, regress `Fare` on:
    1. `nameLength + Age + Parents/Children Aboard `
    2. `nameLength + Age + nameLength*Age`
    3. `log(nameLength) + Age + Parents/Children Aboard`, you will need to use `np.log()`
    4. `Age + Age^2`, you will need to create Age squared
    5. `Age + Age^2 + Age^3`, you will need to create Age cubed

 
- For each regression add below a cell in which you will:
    1. Interpret the R-squared, `Markdown`
    2. Compute the mean squarred error of in-sample prediction using your model, you will first need to use the [predict method](https://www.statsmodels.org/stable/examples/notebooks/generated/predict.html#:~:text=In-sample-,prediction,-%C2%B6) and then use the [mean_squarred_error function](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html) from Scikit-Learn
    
    
- Finally, discuss which model seems to best predict `Fare`. Mention the criteria you used

### Model 1

In [9]:
mod5 = smf.ols('Fare ~ nameLength + Age + Family_Aboard', data=df).fit()
print(mod5.summary())

                            OLS Regression Results                            
Dep. Variable:                   Fare   R-squared:                       0.081
Model:                            OLS   Adj. R-squared:                  0.078
Method:                 Least Squares   F-statistic:                     26.08
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           3.61e-16
Time:                        19:37:04   Log-Likelihood:                -4686.5
No. Observations:                 887   AIC:                             9381.
Df Residuals:                     883   BIC:                             9400.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept        -3.1342      5.736     -0.546

Interpretation

$R^2 = 0.081$ means that ~8% of the variability in Fare is explained by this model (this is low).

In [10]:
Y_pred = mod5.predict(df[['nameLength', 'Age', 'Family_Aboard']])
Y_true = df['Fare'].values
mse = mean_squared_error(Y_true, Y_pred)
print(f"The in-sample MSE of this model is: {mse}")

The in-sample MSE of this model is: 2273.973827424314


### Model 2

In [11]:
mod6 = smf.ols('Fare ~ nameLength + Age + nameLength*Age', data=df).fit()
print(mod6.summary())

                            OLS Regression Results                            
Dep. Variable:                   Fare   R-squared:                       0.040
Model:                            OLS   Adj. R-squared:                  0.037
Method:                 Least Squares   F-statistic:                     12.40
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           6.01e-08
Time:                        19:37:04   Log-Likelihood:                -4705.9
No. Observations:                 887   AIC:                             9420.
Df Residuals:                     883   BIC:                             9439.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          5.0477     13.759      0.

Interpretation

$R^2 = 0.04$ means that ~4% of the variability in Fare is explained by this model (this is low).

In [12]:
Y_pred = mod6.predict(df[['nameLength', 'Age']])
Y_true = df['Fare'].values
mse = mean_squared_error(Y_true, Y_pred)
print(f"The in-sample MSE of this model is: {mse}")

The in-sample MSE of this model is: 2375.399671188585


### Model 3

In [13]:
df['ln_nameLength'] = np.log(df['nameLength'])
 
mod7 = smf.ols('Fare ~ ln_nameLength + Age + Family_Aboard', data=df).fit()
print(mod7.summary())

                            OLS Regression Results                            
Dep. Variable:                   Fare   R-squared:                       0.080
Model:                            OLS   Adj. R-squared:                  0.077
Method:                 Least Squares   F-statistic:                     25.74
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           5.75e-16
Time:                        19:37:04   Log-Likelihood:                -4687.0
No. Observations:                 887   AIC:                             9382.
Df Residuals:                     883   BIC:                             9401.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept       -38.8502     16.759     -2.318

Interpretation

$R^2 = 0.080$ means that ~8% of the variability in Fare is explained by this model (this is low).

In [14]:
Y_pred = mod7.predict(df[['ln_nameLength', 'Age', 'Family_Aboard']])
Y_true = df['Fare'].values
mse = mean_squared_error(Y_true, Y_pred)
print(f"The in-sample MSE of this model is: {mse}")

The in-sample MSE of this model is: 2276.3992879586604


### Model 4

In [15]:
df['Age_2']=np.power(df['Age'],2)
df['Age_3']=np.power(df['Age'],3)

mod8 = smf.ols('Fare ~ Age + Age_2', data=df).fit()
print(mod8.summary())

                            OLS Regression Results                            
Dep. Variable:                   Fare   R-squared:                       0.013
Model:                            OLS   Adj. R-squared:                  0.011
Method:                 Least Squares   F-statistic:                     6.034
Date:                Sun, 07 Nov 2021   Prob (F-statistic):            0.00250
Time:                        19:37:04   Log-Likelihood:                -4718.2
No. Observations:                 887   AIC:                             9442.
Df Residuals:                     884   BIC:                             9457.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     24.9456      6.263      3.983      0.0

Interpretation

$R^2 = 0.013$ means that ~1% of the variability in Fare is explained by our model (this is low).

In [16]:
Y_pred = mod8.predict(df[['Age', 'Age_2']])
Y_true = df['Fare'].values
mse = mean_squared_error(Y_true, Y_pred)
print(f"The in-sample MSE of this model is: {mse}")

The in-sample MSE of this model is: 2442.120741659226


### Model 5

In [17]:
mod9 = smf.ols('Fare ~ Age + Age_2 + Age_3', data=df).fit()
print(mod9.summary())

                            OLS Regression Results                            
Dep. Variable:                   Fare   R-squared:                       0.023
Model:                            OLS   Adj. R-squared:                  0.020
Method:                 Least Squares   F-statistic:                     6.944
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           0.000127
Time:                        19:37:04   Log-Likelihood:                -4713.8
No. Observations:                 887   AIC:                             9436.
Df Residuals:                     883   BIC:                             9455.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     39.4081      7.940      4.963      0.0

Interpretation

$R^2 = 0.023$ means that ~2% of the variability in Fare is explained by our model (this is low).

In [18]:
Y_pred = mod9.predict(df[['Age', 'Age_2', 'Age_3']])
Y_true = df['Fare'].values
mse = mean_squared_error(Y_true, Y_pred)
print(f"The in-sample MSE of this model is: {mse}")

The in-sample MSE of this model is: 2418.4044509511514


### Which model best predicts `Fare`?

All of the models have low predictive ability overall. The model that best predicts Fare is model 1 with variables Age, nameLength, and Family_Aboard. Its R-squared value is the highest (0.081) and its MSE is lowest (~2273). 

## Part 3 - Classification

- Using the pacakge [Statsmodels](https://www.statsmodels.org/stable/example_formulas.html) and OLS regression (linear probability model), regress `Survived` on `Pclass  + Sex + Age + Fare + nameLength`


- Repeat the operation using the [Probit model](https://www.statsmodels.org/stable/generated/statsmodels.formula.api.probit.html)


- Note that you will need to create dummy variables for `Pclass` and `Sex` using Pandas [get_dummies](https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html) function. You only need to do this once for both models
   
   
- Compare, in a markdown cell, the mean absolute errors of both models

In [19]:
# Create dummies

dummies = pd.get_dummies(df[['Pclass', 'Sex']], drop_first=True)
df2 = df.join(dummies)
df2.head()

Survived Pclass                                               Name     Sex  \
0         0      3                             Mr. Owen Harris Braund    male   
1         1      1  Mrs. John Bradley (Florence Briggs Thayer) Cum...  female   
2         1      3                              Miss. Laina Heikkinen  female   
3         1      1        Mrs. Jacques Heath (Lily May Peel) Futrelle  female   
4         0      3                            Mr. William Henry Allen    male   

    Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  \
0  22.0                        1                        0   7.2500   
1  38.0                        1                        0  71.2833   
2  26.0                        0                        0   7.9250   
3  35.0                        1                        0  53.1000   
4  35.0                        0                        0   8.0500   

   nameLength  Family_Aboard  ln_nameLength   Age_2    Age_3  Pclass_2  \
0          22              0       3.091042   484.0  10648.0         0   
1          50              0       3.912023  1444.0  54872.0         0   
2          21              0       3.044522   676.0  17576.0         0   
3          43              0       3.761200  1225.0  42875.0         0   
4          23              0       3.135494  1225.0  42875.0         0   

   Pclass_3  Sex_male  
0         1         1  
1         0         0  
2         1         0  
3         0         0  
4         1         1

### Linear Probability Model

In [20]:
mod10 = smf.ols('Survived ~ Pclass_2 + Pclass_3 + Sex_male + Age + Fare + nameLength', data=df2).fit() 
print(mod10.summary())

                            OLS Regression Results                            
Dep. Variable:               Survived   R-squared:                       0.384
Model:                            OLS   Adj. R-squared:                  0.379
Method:                 Least Squares   F-statistic:                     91.24
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           5.33e-89
Time:                        19:37:04   Log-Likelihood:                -405.39
No. Observations:                 887   AIC:                             824.8
Df Residuals:                     880   BIC:                             858.3
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.0652      0.078     13.594      0.0

In [21]:
Y_pred = mod10.predict(df2[['Pclass_2','Pclass_3','Sex_male','Age','Fare','nameLength']])
Y_true = df2['Survived'].values
mae = mean_absolute_error(Y_true, Y_pred)
print(f"The in-sample MAE of this model is: {mae}")

The in-sample MAE of this model is: 0.2951239821172641


### Logit (as per email)

In [22]:
mod11 = smf.logit('Survived ~ Pclass_2 + Pclass_3 + Sex_male + Age + Fare + nameLength', data=df2).fit()
print(mod11.summary())

Optimization terminated successfully.
         Current function value: 0.451569
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:               Survived   No. Observations:                  887
Model:                          Logit   Df Residuals:                      880
Method:                           MLE   Df Model:                            6
Date:                Sun, 07 Nov 2021   Pseudo R-squ.:                  0.3227
Time:                        19:37:04   Log-Likelihood:                -400.54
converged:                       True   LL-Null:                       -591.38
Covariance Type:            nonrobust   LLR p-value:                 2.415e-79
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      3.3666      0.538      6.263      0.000       2.313       4.420
Pclass_2      -1.1705      0.

In [23]:
Y_pred = mod11.predict(df2[['Pclass_2','Pclass_3','Sex_male','Age','Fare','nameLength']])
Y_true = df2['Survived'].values
mae = mean_absolute_error(Y_true, Y_pred)
print(f"The in-sample MAE of this model is: {mae}")

The in-sample MAE of this model is: 0.2878781039135286


### Compare the mean absolute errors of both models

The MAE for the logistic regression model is smaller than the MAE for the linear probability model (0.288 vs. 0.452).  The logit model is a much better predictor for Survived. 